### PR TITLE
Backport pvc resiliency to 0.7

### LINKED ISSUE #2072: Rook Operator Exits Creating PVC
**Is this a bug report or feature request?**
Bug Report

**Deviation from expected behavior:**
Rook operator exits in creation of PVC and then enters loop of unable to create PVC.
**Expected behavior:**
PVC created properly.
**How to reproduce it (minimal and precise):**
Create numerous PVC (>5) at the same time (in a single kubectl apply) on a rook cluster
**Environment**:
* OS (e.g. from /etc/os-release): Ubuntu 16.04
* Kernel (e.g. `uname -a`):
Linux cnat-srv-1 4.4.0-87-generic #110-Ubuntu SMP Tue Jul 18 12:55:35 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
* Cloud provider or hardware configuration: kubeadm
* Rook version (use `rook version` inside of a Rook Pod): 0.8.1
* Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T20:17:28Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T20:08:34Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): kubeadm
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/Rook/master/toolbox.html)): healthy

```
2018-08-27 15:09:57.267869 I | op-provisioner: creating volume with configuration {pool:replicapool clusterNamespace:rook-ceph fstype:}
2018-08-27 15:09:57.267932 I | exec: Running command: rbd create replicapool/pvc-42775829-aa0b-11e8-8620-0025b500257e --size 3072 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring

2018-08-27 15:09:59.050598 I | op-provisioner: creating volume with configuration {pool:replicapool clusterNamespace:rook-ceph fstype:}
2018-08-27 15:09:59.050640 I | exec: Running command: rbd create replicapool/pvc-433c9bb6-aa0b-11e8-8620-0025b500257e --size 1024 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring
2018-08-27 15:09:59.450261 I | op-provisioner: creating volume with configuration {pool:replicapool clusterNamespace:rook-ceph fstype:}
2018-08-27 15:09:59.450299 I | exec: Running command: rbd create replicapool/pvc-43480416-aa0b-11e8-8620-0025b500257e --size 5120 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring
2018-08-27 15:10:00.475865 I | op-provisioner: creating volume with configuration {pool:replicapool clusterNamespace:rook-ceph fstype:}
2018-08-27 15:10:00.475916 I | exec: Running command: rbd create replicapool/pvc-4352c43f-aa0b-11e8-8620-0025b500257e --size 5120 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring
E0827 15:10:00.950328       7 leaderelection.go:273] Failed to update lock: Operation cannot be fulfilled on persistentvolumeclaims "db-db-admin-1": the object has been modified; please apply your changes to the latest version and try again
2018-08-27 15:10:04.550831 I | exec: Running command: rbd ls -l replicapool --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json
2018-08-27 15:10:05.450224 I | op-provisioner: creating volume with configuration {pool:replicapool clusterNamespace:rook-ceph fstype:}
2018-08-27 15:10:05.450259 I | exec: Running command: rbd create replicapool/pvc-436f33d2-aa0b-11e8-8620-0025b500257e --size 5120 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring
2018-08-27 15:10:07.453563 I | exec: Running command: rbd ls -l replicapool --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json
2018-08-27 15:10:08.651269 I | exec: Running command: rbd ls -l replicapool --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json
2018-08-27 15:10:09.750128 I | exec: Running command: rbd ls -l replicapool --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json
2018-08-27 15:10:12.653163 I | exec: Running command: rbd ls -l replicapool --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json
2018-08-27 15:10:23.150449 I | operator: shutdown signal received, exiting...
```
When it restarts the following look occurs:
```
E0827 15:11:55.550288       7 controller.go:801] Failed to provision volume for claim "app1/db-db-admin-1" with StorageClass "rook-ceph-block": Failed to create rook block image replicapool/pvc-436f33d2-aa0b-11e8-8620-0025b500257e: failed to create image pvc-436f33d2-aa0b-11e8-8620-0025b500257e in pool replicapool of size 5368709120: Failed to complete '': exit status 17. rbd: create error: (17) File exists
2018-08-27 15:11:54.551573 7f88fd477d40 -1 librbd: rbd image pvc-436f33d2-aa0b-11e8-8620-0025b500257e already exists
. output: 2018-08-27 15:11:54.551573 7f88fd477d40 -1 librbd: rbd image pvc-436f33d2-aa0b-11e8-8620-0025b500257e already exists
E0827 15:11:55.550404       7 goroutinemap.go:165] Operation for "provision-app1/db-db-admin-1[436f33d2-aa0b-11e8-8620-0025b500257e]" failed. No retries permitted until 2018-08-27 15:12:03.55037151 +0000 UTC m=+97.397060281 (durationBeforeRetry 8s). Error: Failed to create rook block image replicapool/pvc-436f33d2-aa0b-11e8-8620-0025b500257e: failed to create image pvc-436f33d2-aa0b-11e8-8620-0025b500257e in pool replicapool of size 5368709120: Failed to complete '': exit status 17. rbd: create error: (17) File exists
2018-08-27 15:11:54.551573 7f88fd477d40 -1 librbd: rbd image pvc-436f33d2-aa0b-11e8-8620-0025b500257e already exists
. output: 2018-08-27 15:11:54.551573 7f88fd477d40 -1 librbd: rbd image pvc-436f33d2-aa0b-11e8-8620-0025b500257e already exists
E0827 15:11:55.650241       7 controller.go:801] Failed to provision volume for claim "app1/pb-repositories-app1-app1-engine-app-baseline" with StorageClass "rook-ceph-block": Failed to create rook block image replicapool/pvc-433c9bb6-aa0b-11e8-8620-0025b500257e: failed to create image pvc-433c9bb6-aa0b-11e8-8620-0025b500257e in pool replicapool of size 1073741824: Failed to complete '': exit status 17. rbd: create error: (17) File exists
2018-08-27 15:11:54.950923 7f9b1e8d1d40 -1 librbd: rbd image pvc-433c9bb6-aa0b-11e8-8620-0025b500257e already exists
. output: 2018-08-27 15:11:54.950923 7f9b1e8d1d40 -1 librbd: rbd image pvc-433c9bb6-aa0b-11e8-8620-0025b500257e already exists
E0827 15:11:55.650358       7 goroutinemap.go:165] Operation for "provision-app1/pb-repositories-app1-app1-engine-app-baseline[433c9bb6-aa0b-11e8-8620-0025b500257e]" failed. No retries permitted until 2018-08-27 15:12:03.650319064 +0000 UTC m=+97.497007801 (durationBeforeRetry 8s). Error: Failed to create rook block image replicapool/pvc-433c9bb6-aa0b-11e8-8620-0025b500257e: failed to create image pvc-433c9bb6-aa0b-11e8-8620-0025b500257e in pool replicapool of size 1073741824: Failed to complete '': exit status 17. rbd: create error: (17) File exists
2018-08-27 15:11:54.950923 7f9b1e8d1d40 -1 librbd: rbd image pvc-433c9bb6-aa0b-11e8-8620-0025b500257e already exists
. output: 2018-08-27 15:11:54.950923 7f9b1e8d1d40 -1 librbd: rbd image pvc-433c9bb6-aa0b-11e8-8620-0025b500257e already exists
E0827 15:11:55.750112       7 controller.go:801] Failed to provision volume for claim "app1/db-db-session-config-1" with StorageClass "rook-ceph-block": Failed to create rook block image replicapool/pvc-4352c43f-aa0b-11e8-8620-0025b500257e: failed to create image pvc-4352c43f-aa0b-11e8-8620-0025b500257e in pool replicapool of size 5368709120: Failed to complete '': exit status 17. rbd: create error: (17) File exists
2018-08-27 15:11:55.050053 7f22ee6f8d40 -1 librbd: rbd image pvc-4352c43f-aa0b-11e8-8620-0025b500257e already exists
. output: 2018-08-27 15:11:55.050053 7f22ee6f8d40 -1 librbd: rbd image pvc-4352c43f-aa0b-11e8-8620-0025b500257e already exists
E0827 15:11:55.750203       7 goroutinemap.go:165] Operation for "provision-app1/db-db-session-config-1[4352c43f-aa0b-11e8-8620-0025b500257e]" failed. No retries permitted until 2018-08-27 15:12:03.75017653 +0000 UTC m=+97.596865257 (durationBeforeRetry 8s). Error: Failed to create rook block image replicapool/pvc-4352c43f-aa0b-11e8-8620-0025b500257e: failed to create image pvc-4352c43f-aa0b-11e8-8620-0025b500257e in pool replicapool of size 5368709120: Failed to complete '': exit status 17. rbd: create error: (17) File exists
2018-08-27 15:11:55.050053 7f22ee6f8d40 -1 librbd: rbd image pvc-4352c43f-aa0b-11e8-8620-0025b500257e already exists
. output: 2018-08-27 15:11:55.050053 7f22ee6f8d40 -1 librbd: rbd image pvc-4352c43f-aa0b-11e8-8620-0025b500257e already exists
E0827 15:11:55.953280       7 controller.go:801] Failed to provision volume for claim "app1/db-db-session-config-0" with StorageClass "rook-ceph-block": Failed to create rook block image replicapool/pvc-43480416-aa0b-11e8-8620-0025b500257e: failed to create image pvc-43480416-aa0b-11e8-8620-0025b500257e in pool replicapool of size 5368709120: Failed to complete '': exit status 17. rbd: create error: (17) File exists
2018-08-27 15:11:54.750203 7fac2ce34d40 -1 librbd: rbd image pvc-43480416-aa0b-11e8-8620-0025b500257e already exists
. output: 2018-08-27 15:11:54.750203 7fac2ce34d40 -1 librbd: rbd image pvc-43480416-aa0b-11e8-8620-0025b500257e already exists
E0827 15:11:55.953400       7 goroutinemap.go:165] Operation for "provision-app1/db-db-session-config-0[43480416-aa0b-11e8-8620-0025b500257e]" failed. No retries permitted until 2018-08-27 15:12:03.9533663 +0000 UTC m=+97.800055032 (durationBeforeRetry 8s). Error: Failed to create rook block image replicapool/pvc-43480416-aa0b-11e8-8620-0025b500257e: failed to create image pvc-43480416-aa0b-11e8-8620-0025b500257e in pool replicapool of size 5368709120: Failed to complete '': exit status 17. rbd: create error: (17) File exists
2018-08-27 15:11:54.750203 7fac2ce34d40 -1 librbd: rbd image pvc-43480416-aa0b-11e8-8620-0025b500257e already exists
. output: 2018-08-27 15:11:54.750203 7fac2ce34d40 -1 librbd: rbd image pvc-43480416-aa0b-11e8-8620-0025b500257e already exists
E0827 15:11:56.156070       7 controller.go:801] Failed to provision volume for claim "app1/svn-db" with StorageClass "rook-ceph-block": Failed to create rook block image replicapool/pvc-42775829-aa0b-11e8-8620-0025b500257e: failed to create image pvc-42775829-aa0b-11e8-8620-0025b500257e in pool replicapool of size 3221225472: Failed to complete '': exit status 17. rbd: create error: (17) File exists
2018-08-27 15:11:55.850190 7f61c8ad5d40 -1 librbd: rbd image pvc-42775829-aa0b-11e8-8620-0025b500257e already exists
. output: 2018-08-27 15:11:55.850190 7f61c8ad5d40 -1 librbd: rbd image pvc-42775829-aa0b-11e8-8620-0025b500257e already exists
E0827 15:11:56.156171       7 goroutinemap.go:165] Operation for "provision-app1/svn-db[42775829-aa0b-11e8-8620-0025b500257e]" failed. No retries permitted until 2018-08-27 15:12:04.156142499 +0000 UTC m=+98.002831222 (durationBeforeRetry 8s). Error: Failed to create rook block image replicapool/pvc-42775829-aa0b-11e8-8620-0025b500257e: failed to create image pvc-42775829-aa0b-11e8-8620-0025b500257e in pool replicapool of size 3221225472: Failed to complete '': exit status 17. rbd: create error: (17) File exists
```
This problem seems to be related to rapid creation of PVC when charts like DBs are deployed ( in this case mongo).

### PR DESCRIPTION
**Pull request by clkao**
**Description of your changes:**

Backport #2108 1ce5a4d7 and 8b3b73b3 to 0.7 release branch

**Which issue is resolved by this Pull Request:**
Resolves #2072

### CODE REVIEW

File: pkg/daemon/ceph/client/image.go [hunk 1]
```diff
@@ -75,8 +77,14 @@ func CreateImage(context *clusterd.Context, clusterName, name, poolName string,
 	args := []string{"create", imageSpec, "--size", strconv.Itoa(sizeMB)}
 	buf, err := ExecuteRBDCommandNoFormat(context, clusterName, args)
 	if err != nil {
-		return nil, fmt.Errorf("failed to create image %s in pool %s of size %d: %+v. output: %s",
-			name, poolName, size, err, string(buf))
+		cmdErr, ok := err.(*exec.CommandError)
+		if ok && cmdErr.ExitStatus() == int(syscall.EEXIST) {
+			// Image with the same name already exists in the given rbd pool. Continuing with the link to PV.
+			logger.Warningf("Requested image %s exists in pool %s. Continuing", name, poolName)
```
Comment (arm64b) [inline pkg/daemon/ceph/client/image.go]: I am wording if we really need this, with this change, meaning we will have 2 exactly the same (pv:volume) mapping, but each has a different PVC object, is that the right thing for the storage consuming?

**Comment by travisn** [comment]
@clkao we are not planning on 0.7 releases since the 0.8 release. Could you upgrade to 0.8? Note that this fix will be included in 0.8.2 hopefully in the next few days, so you might wait for that. Thanks.

**Comment by clkao** [comment]
@travisn (continuing the discussion from https://github.com/rook/rook/pull/2108#issuecomment-421072033)

When the issue manifests, the cluster is in an inconsistent state where PVCs are pending and operator might be restarting. In the upgrade guide, we do ask the users to verify all pods using the rook volumes should be healthily running, but in this case it's likely the pods consuming rook-backed PVCs are in pending state as the PVCs in question are pending.

As the 0.8 upgrade procedure is rather manual and involving, I'd suggest we give users sufficient confidence to upgrade from a consistent state of a 0.7.x release.

**Comment by ganeshmaharaj** [comment]
@clkao v0.8.2 with this fix has been released. i believe upgrading to that version should fix this error. there are two ways. 1. you can delete the failing pvcs, upgrade the operator and then re-create them. 2. upgrade operator and restart it, which will get to this point eventually and be able to link them properly. @travisn @jbw976 @galexrt am i right in those statements?

@clkao if that works, can we close this PR?

**Comment by clkao** [comment]
@ganeshmaharaj as the current [upgrade doc](https://rook.io/docs/rook/v0.8/upgrade.html) states:

> All pods consuming Rook storage should be created, running, and in a steady state. No Rook persistent volumes should be in the act of being created or deleted.

Should we add a note about the pending pvc can be safely deleted in case people run into this issue, and then they can proceed to upgrade?

**Comment by ganeshmaharaj** [comment]
@clkao PR https://github.com/rook/rook/pull/2156 should address your concern.

**Comment by travisn** [comment]
@clkao I would expect the PVCs will stay in the pending state through the upgrade and then after the upgrade, the issue would be fixed as @ganeshmaharaj mentioned. The comment in the upgrade guide about not creating PVCs during upgrade is simply to point out that you want to minimize the operations in the cluster during the upgrade. Have you tried the upgrade yet?

**Comment by clkao** [comment]
Since we increased resource limits, we have not seen this operator-crashing error that requires the fix.

**Pull request closed without merging**